overlord/snapstate/handlers.go

```diff
@@ -1147,9 +1149,43 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		}
 		return copyDataErr
 	}
+
+	// the migration hasn't been done - do it now
+	if opts.UseHidden && !opts.MigratedToHidden {
+		if err := m.migrateToHiddenDir(t); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
+func (m *SnapManager) migrateToHiddenDir(t *state.Task) error {
```

Comment:
this could probably take the same paramters as undoMigrateToHidden and you can skip TaskSnapSetup then